RecoEgamma/EgammaTools/src/EgammaPCAHelper.cc

```diff
@@ -0,0 +1,342 @@
+            std::cout << rh_detid.det() << " " << HGCalDetId(rh_detid) << std::endl;
+            continue;
+        }
+        if (debug_) {
+            std::cout << "DetId " << rh_detid.rawId() << " " << layer << " " <<  itcheck->second->energy() <<std::endl;
+            //        std::cout << " Hit " << itcheck->second << " " << itcheck->second->energy() << std::endl;
+        }
+        float fraction = hf[j].second;
+
+        double thickness =
+            (DetId::Forward == DetId(rh_detid).det()) ? recHitTools_->getSiThickness(rh_detid) : -1;
+        double mip = dEdXWeights_[layer] * 0.001;  // convert in GeV
+        if (thickness > 99. && thickness < 101)
+            mip *= invThicknessCorrection_[0];
+        else if (thickness > 199 && thickness < 201)
+            mip *= invThicknessCorrection_[1];
+        else if (thickness > 299 && thickness < 301)
+            mip *= invThicknessCorrection_[2];
+
+        pcavars[0] = recHitTools_->getPosition(rh_detid).x();
+        pcavars[1] = recHitTools_->getPosition(rh_detid).y();
+        pcavars[2] = recHitTools_->getPosition(rh_detid).z();
+        if (pcavars[2] == 0.)
+            std::cout << " Problem, hit with z =0 ";
+        else  {
+            Spot mySpot(rh_detid,itcheck->second->energy(),pcavars,layer,fraction,mip);
+            theSpots_.push_back(mySpot);
+        }
+    }
+    if (debug_) {
+        std::cout << " Stored " << theSpots_.size() << " hits " << std::endl;
+    }
+}
+
+void EGammaPCAHelper::computePCA(float radius , bool withHalo) {
+    // very important - to reset
+    pca_.reset(new TPrincipal(3, "D"));
+    bool initialCalculation = radius < 0;
+    if (debug_)
+        std::cout << " Initial calculation " << initialCalculation << std::endl;
+    if (initialCalculation && withHalo) {
+        std::cout << "Warning - in the first iteration, the halo hits are excluded " << std::endl;
+        withHalo=false;
+    }
+
+    float radius2 = radius*radius;
+    if (! initialCalculation)     {
+        math::XYZVector mainAxis(axis_);
+        mainAxis.unit();
+        math::XYZVector phiAxis(barycenter_.x(), barycenter_.y(), 0);
+        math::XYZVector udir(mainAxis.Cross(phiAxis));
+        udir = udir.unit();
+        trans_ = Transform3D(Point(barycenter_), Point(barycenter_ + axis_), Point(barycenter_ + udir), Point(0, 0, 0),
+        Point(0., 0., 1.), Point(1., 0., 0.));
+    }
+
+    unsigned nSpots = theSpots_.size();
+    std::set<int> layers;
+    for ( unsigned i =0; i< nSpots ; ++i) {
+        Spot spot(theSpots_[i]);
+        if (spot.layer() > 28) continue;
+        if (!withHalo && (! spot.isCore() ))
+            continue;
+        if (initialCalculation) {
+            // initial calculation, take only core hits
+            if ( ! spot.isCore() ) continue;
+            //std::cout << " Multiplicity " << spot.multiplicity() << " " << spot.row()[0] << " " ;
+            //std::cout << spot.row()[1] << " " << spot.row()[2] << std::endl;
```

Comment:
is this commented out code needed?
Please remove or add comments inline in the code why the commented out block is relevant

If debugs are needed, add them via LogDebug or LogTrace